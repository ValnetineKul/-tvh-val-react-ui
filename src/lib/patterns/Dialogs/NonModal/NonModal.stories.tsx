import type { ComponentProps } from 'react';
import React, { useState } from 'react';
import { makeStyles } from '../../../../themes/core';
import { createMeta } from '../../../story-utils';
import type { StoryTemplate } from '../../../story-utils';
import NonModal from './NonModal';
import Typography from '../../../components/Typography';
import Button from '../../../components/Buttons/Button';

const useStyles = makeStyles()((theme) => ({
  root: {
    width: '320px',
  },
  buttons: {
    display: 'flex',
    marginTop: theme.layoutSpacing['spacing/400'],
    gap: theme.layoutSpacing['spacing/400'],
  },
}));

type Props = ComponentProps<typeof NonModal>;
type Args = Omit<Props, 'isOpen' | 'onClose'>;

const Template = ((args: Args) => {
  const { classes } = useStyles();
  const [isOpen, setIsOpen] = useState(true);
  return (
    <>
      {!isOpen && (
        <Button
          variant="tertiary"
          label="Open Nonmodal"
          onClick={() => {
            setIsOpen(true);
          }}
        />
      )}
      <NonModal
        {...args}
        isOpen={isOpen}
        className={classes.root}
        onClose={() => {
          setIsOpen(false);
        }}
      >
        <Typography>
          We use cookies on this site to enhance your user experience. By clicking any link on this page you are giving
          your consent for us to set cookies.
        </Typography>
        <div className={classes.buttons}>
          <Button
            label="Accept cookies"
            onClick={() => {
              setIsOpen(false);
            }}
          />
          <Button
            variant="tertiary"
            label="Manage cookies"
            onClick={() => {
              setIsOpen(false);
            }}
          />
        </div>
      </NonModal>
    </>
  );
}) as StoryTemplate<Args>;

export const Primitive = Template.bind({});
Primitive.args = {
  priority: 'normal',
  position: 'bottom-start',
  title: 'Cookies',
};

Primitive.storyName = 'NonModal';
export default createMeta({
  component: NonModal,
  title: 'Patterns/Dialogs/NonModal',
  argTypes: {
    isOpen: {
      control: {
        disable: true,
      },
    },
    title: {
      control: {
        disable: true,
      },
    },
    onClose: {
      control: {
        disable: true,
      },
    },
  },
});
