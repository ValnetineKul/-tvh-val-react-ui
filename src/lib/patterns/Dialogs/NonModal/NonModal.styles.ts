import { makeStyles } from '../../../../themes/core';

const useStyles = makeStyles({ name: 'NonModal' })((theme) => ({
  root: {
    position: 'fixed',
    zIndex: 1500,
    boxShadow: theme.shadow['shadow/500'],
    [theme.breakpoints.down('sm')]: {
      width: '100%',
    },
  },
  positionBottomStart: {
    left: theme.layoutSpacing['spacing/000'],
    bottom: theme.layoutSpacing['spacing/450'],
    [theme.breakpoints.down('sm')]: {
      bottom: theme.layoutSpacing['spacing/000'],
    },
  },
  positionBottomEnd: {
    right: theme.layoutSpacing['spacing/000'],
    bottom: theme.layoutSpacing['spacing/450'],
    [theme.breakpoints.down('sm')]: {
      bottom: theme.layoutSpacing['spacing/000'],
    },
  },
  title: {
    padding: `${theme.layoutSpacing['spacing/300']} ${theme.layoutSpacing['spacing/400']}`,
    display: 'flex',
    justifyContent: 'space-between',

    [theme.breakpoints.down('sm')]: {
      padding: `${theme.layoutSpacing['spacing/200']} ${theme.layoutSpacing['spacing/400']}`,
    },
  },
  titleText: {
    whiteSpace: 'break-spaces',
    marginTop: theme.layoutSpacing['spacing/200'],
  },
  children: {
    padding: theme.layoutSpacing['spacing/400'],
  },
  closeButton: {
    alignSelf: 'flex-start',
  },
}));

export default useStyles;
