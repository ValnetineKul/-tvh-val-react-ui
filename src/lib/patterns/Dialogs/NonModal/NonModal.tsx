import type { FC } from 'react';
import React from 'react';
import Slide from '@mui/material/Slide';
import { default as SurfaceComponent } from '../../../components/Surfaces/Surface';
import useStyles from './NonModal.styles';
import IconButton from '../../../components/Buttons/IconButton';
import Icon from '../../../components/Icon';
import { Times } from '../../../components/Icon/icons/functional';
import { ARIA_CLOSE_DIALOG } from '../Modal';
import Typography from '../../../components/Typography';

export interface NonModalProps {
  className?: string;
  isOpen: boolean;
  onClose?: React.MouseEventHandler<HTMLButtonElement>;
  position?: 'bottom-start' | 'bottom-end' | 'unset';
  priority?: 'normal' | 'high';
  title?: string | React.ReactNode;
}

const NonModal: FC<NonModalProps> = ({
  className,
  isOpen,
  onClose,
  position = 'bottom-start',
  priority = 'normal',
  title,
  children,
}) => {
  const { classes, cx } = useStyles();

  const positionClassName = position === 'bottom-start' ? classes.positionBottomStart : classes.positionBottomEnd;

  const renderTitle = () => {
    return typeof title === 'string' ? (
      <Typography variant="h4" className={classes.titleText}>
        {title}
      </Typography>
    ) : (
      title
    );
  };

  const renderTitleRow = () => {
    return (
      <div className={classes.title}>
        <div>{title && renderTitle()}</div>
        {onClose && (
          <IconButton
            className={classes.closeButton}
            size="md"
            onClick={onClose}
            icon={<Icon icon={Times} />}
            aria-label={ARIA_CLOSE_DIALOG}
          />
        )}
      </div>
    );
  };

  return (
    <Slide direction="up" in={isOpen} mountOnEnter unmountOnExit>
      <SurfaceComponent
        role="dialog"
        className={cx(classes.root, { [positionClassName]: position !== 'unset' }, className)}
        color={priority === 'normal' ? '100' : 'Secondary'}
      >
        {(!!title || onClose) && renderTitleRow()}
        <div className={classes.children}>{children}</div>
      </SurfaceComponent>
    </Slide>
  );
};

export default NonModal;
