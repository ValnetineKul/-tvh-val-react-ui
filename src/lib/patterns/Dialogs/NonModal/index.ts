import NonModal from './NonModal';

export type { NonModalProps } from './NonModal';
export default NonModal;
