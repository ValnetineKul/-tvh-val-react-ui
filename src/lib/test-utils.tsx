import userEvent from '@testing-library/user-event';
import { fireEvent, render } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { theme } from '../themes/tvh';
import { createTheme, ThemeProvider, SurfaceProvider } from '../themes/core';

export const muiTheme = createTheme(theme);

const AllTheProviders: React.FC = ({ children }) => {
  return (
    <MemoryRouter>
      <ThemeProvider theme={muiTheme}>
        <SurfaceProvider color="100">{children}</SurfaceProvider>
      </ThemeProvider>
    </MemoryRouter>
  );
};

const customRender = (component: React.ReactElement, options?: Parameters<typeof render>[1]) =>
  render(component, { wrapper: AllTheProviders, ...options });

const customRenderHook: typeof renderHook = (hook, options) =>
  renderHook(hook, { wrapper: AllTheProviders, ...options });

export const initResponsiveTest = (size: number | 'Mobile' | 'Tablet' | 'Desktop' | 'LgDesktop' | 'XlDesktop') => {
  let width: number;
  if (size === 'Mobile') width = muiTheme.breakpoints.values.sm - 1;
  else if (size === 'Tablet') width = muiTheme.breakpoints.values.sm;
  else if (size === 'Desktop') width = muiTheme.breakpoints.values.md;
  else if (size === 'LgDesktop') width = muiTheme.breakpoints.values.lg;
  else if (size === 'XlDesktop') width = muiTheme.breakpoints.values.xl;
  else width = size;

  Object.defineProperty(window, 'innerWidth', {
    writable: true,
    configurable: true,
    value: width,
  });

  window.matchMedia = jest.fn().mockImplementation((query) => {
    const queryWidthInterval: string[] = query.match(/\d+(\.\d+)?/g);

    let matches;
    if (queryWidthInterval.length === 2) {
      matches = width >= +queryWidthInterval[0] && width < +queryWidthInterval[1];
    } else {
      const queryWidth = +queryWidthInterval[0];
      const isDown = /max/.test(query);

      matches = isDown ? width <= queryWidth : width >= queryWidth;
    }

    return {
      matches,
      media: query,
      onchange: null,
      addListener: jest.fn(),
      removeListener: jest.fn(),
    };
  });
};

const customUserEvent = Object.assign(userEvent, {
  mouseDown: fireEvent.mouseDown,
  mouseUp: fireEvent.mouseUp,
  mouseMove: fireEvent.mouseMove,
  scroll: fireEvent.scroll,
  error: fireEvent.error,
});

export { screen, within, waitFor, act, waitForElementToBeRemoved } from '@testing-library/react';
// override render method
export { customRender as render, customRenderHook as renderHook, customUserEvent as userEvent };
