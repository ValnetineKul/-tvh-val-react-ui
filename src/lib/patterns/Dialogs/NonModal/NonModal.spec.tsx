import type { ComponentProps } from 'react';
import React from 'react';
import { screen, render, userEvent } from '../../../test-utils';
import NonModal from './NonModal';

type Props = ComponentProps<typeof NonModal>;

const props: Props = {
  isOpen: true,
  title: 'Title',
  onClose: jest.fn(),
};

describe('NonModal', () => {
  it('Should render correctly', () => {
    const { baseElement } = render(<NonModal {...props}>Content</NonModal>);
    expect(baseElement).toMatchSnapshot();
  });

  it('Should not render title', () => {
    const { title, ...noTitleProps } = props;
    render(
      <NonModal {...noTitleProps} priority="high">
        Content
      </NonModal>
    );

    const titleElement = screen.queryByText(/Title/);

    expect(titleElement).not.toBeInTheDocument();
  });

  describe('Should set position className', () => {
    const cases: [Props['position'], string][] = [
      [undefined, 'NonModal-positionBottomStart'],
      ['bottom-start', 'NonModal-positionBottomStart'],
      ['bottom-end', 'NonModal-positionBottomEnd'],
    ];
    test.each(cases)('given %p, returns %p', (input, expected) => {
      render(
        <NonModal position={input} isOpen>
          content
        </NonModal>
      );
      const content = screen.getByText('content');
      expect(content?.parentElement?.className).toMatch(expected);
    });
  });

  it('Should callback on header close button click', () => {
    render(<NonModal {...props} />);
    userEvent.click(screen.getByRole('button'));
    expect(props.onClose).toHaveBeenCalledTimes(1);
  });
});
