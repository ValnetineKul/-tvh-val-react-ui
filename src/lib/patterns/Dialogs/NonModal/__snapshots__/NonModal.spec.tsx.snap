// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NonModal Should render correctly 1`] = `
<body>
  <div>
    <div
      class="tss-98rpvu-Surface-root-NonModal-root-NonModal-positionBottomStart"
      role="dialog"
      style="webkit-transform: none; transform: none; webkit-transition: -webkit-transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms; transition: transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;"
    >
      <div
        class="tss-9x5nbr-NonModal-title"
      >
        <div>
          <h4
            class="tss-4-Typography-headerFunctional-ref tss-ptoxjj-Typography-root-Typography-headerFunctional-NonModal-titleText-Typography-h4"
          >
            Title
          </h4>
        </div>
        <button
          aria-label="Close dialog"
          class="MuiButtonBase-root tss-s1ihq7-ButtonBase-root-ButtonBase-focusOut-IconButton-root-NonModal-closeButton mui-10d1a0h-MuiButtonBase-root"
          tabindex="0"
          type="button"
        >
          <div
            class="tss-w1sjks-Icon-root-Icon-sizeMd"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium mui-i4bv87-MuiSvgIcon-root"
              color="inherit"
              focusable="false"
              viewBox="0 0 24 24"
            >
              times.svg
            </svg>
          </div>
        </button>
      </div>
      <div
        class="tss-1ub92gz-NonModal-children"
      >
        Content
      </div>
    </div>
  </div>
</body>
`;
